psl-core/src/main/java/org/linqs/psl/reasoner/dcd/DCDReasoner.java

```diff
@@ -55,67 +56,79 @@ public double optimize(TermStore baseTermStore) {
         }
 
         @SuppressWarnings("unchecked")
-        VariableTermStore<DCDObjectiveTerm, RandomVariableAtom> termStore = (VariableTermStore<DCDObjectiveTerm, RandomVariableAtom>)baseTermStore;
+        VariableTermStore<DCDObjectiveTerm, GroundAtom> termStore = (VariableTermStore<DCDObjectiveTerm, GroundAtom>)baseTermStore;
 
         termStore.initForOptimization();
 
-        // This must be called after the term store has to correct variable capacity.
-        // A reallocation can cause this array to become out-of-date.
-        float[] variableValues = termStore.getVariableValues();
+        long termCount = 0;
+        double objective = Double.POSITIVE_INFINITY;
 
-        double objective = -1.0;
+        // Starting the second round of iteration, keep track of the old objective.
+        // Note that the number of variables may change in the first iteration (since grounding happens then).
         double oldObjective = Double.POSITIVE_INFINITY;
-
-        if (printInitialObj && log.isTraceEnabled()) {
-            objective = computeObjective(termStore, variableValues);
-            log.trace("Iteration {} -- Objective: {}, Iteration Time: {}, Total Optimiztion Time: {}", 0, objective, 0, 0);
-        }
+        float[] oldVariableValues = null;
 
         int iteration = 1;
         long totalTime = 0;
         while (true) {
             long start = System.currentTimeMillis();
 
+            termCount = 0;
+            objective = 0.0;
+
             for (DCDObjectiveTerm term : termStore) {
-                term.minimize(truncateEveryStep, variableValues);
+                if (oldVariableValues != null) {
+                    objective += term.evaluate(oldVariableValues);
```

Comment:
In `computeObjective`, this is divided by `c`.
I don't know which one is right off the top of my head, but I imagine they should be consistent.